Basic multiturn irrigation

KulkoBSW opened the pull request:
@TomWerner 
I have created a first version and am looking for a bit of feedback before diving deeper in,

What it does for now:
- Make irrigation take 4 turns
- Set the worker busy while working
- - Multiple workers working on the same project

What it does not for now:
- animations
- use values from Terraform (instead every number is hardcoded variable)
- Difference between workers and slaves.
- Different update rules for Multiplayer
- Saving and loading

What I am not sure about:
I update the progress on the end of the current turn (Which seems to fit with the functionality described in the doc). But you handle other similar things (healing) in OnBeginTurn. 
How would I even begin to make a different behavior Multiplayer (update each Worker individually) vs. Singleplayer (Update all workers at one point and then check if the Job is done)

lt-brlara committed:
chore: :memo: upgrade target .NET framework to 8.0

This change modifies all .csproj to target .NET 8.0.

This change also upgrades to the latest Godot SDK version of 4.3.0.

lt-brlara committed:
ci: :construction_worker: Upgrade from .net7.0 to .net8.0

TomWerner committed:
Support turning citizens into specialists and changing their type

This doesn't yet implement the benefits of specialists, but it does allow us to force certain tiles to not be worked. This is important when passing tiles between cities, for example.

#541

TomWerner committed:
Provide tile assignments for cities in scenarios

Unlike with `.SAV` files, scenarios don't specify the tile assignments for each of the citizens in cities. To handle this we backfill them using the tile assignment AI. This is slightly complicated by fact that the `C7Engine` code can't depend directly on the AI logic, so we have to do some dependency injection via an `Action`.

#541

TomWerner committed:
Load culture values from saves and scenarios and display them

This doesn't yet handle per-turn culture (we'll need building for that) but it should allow us to determine the proper border expansion levels in a following PR, as well as improve the tile assignment logic to restrict things to the BFC.

#541

TomWerner committed:
Properly calculate city borders based on culture

We now have nearly tile-for-tile accuracy with the real game for loaded save files! 

I've based this mainly off of https://forums.civfanatics.com/threads/the-eight-laws-of-border-dynamics.106882/, but haven't yet implemented rules 7 or 8

KulkoBSW committed:
Irrigating sets progress and checks if the Job is finished

KulkoBSW committed:
Irrigating sets progress and checks if the Job is finished

KulkoBSW committed:
Added End of Turn Handling

KulkoBSW committed:
Added End of Turn Handling

TomWerner commented on C7GameData/Tile.cs:
```diff
@@ -251,6 +251,7 @@ public static Tuple<int, int> NeighborCoordinate(int X, int Y, TileDirection dir
 			}
 			return Tuple.Create(X, Y);
 		}
+
```
could you revert this extra line?

TomWerner commented on C7Engine/EntryPoints/TurnHandling.cs:
```diff
@@ -69,6 +70,7 @@ internal static void AdvanceTurn() {
 			}
 		}
 
+
```
could you revert this extra line?

TomWerner commented on C7Engine/TileExtensions.cs:
```diff
@@ -40,6 +41,91 @@ public static void DisbandNonDefendingUnits(this Tile tile) {
+				if (currentWorkerJob.Equals(unit.WorkerJob)) {
+					totalProgress += unit.WorkerProgressTowardsJob;
+				} else {
+					// reset Unit working on other jobs
```
This needs to be reset if the worker is selected/moved while doing the work - otherwise you can start irrigating on a legal tile, then move to a different tile and let the irrigation finish.

Here I managed to irrigate a hill

![image](https://github.com/user-attachments/assets/c801eb3e-c426-48b1-b37a-eee9e408fd2b)

KulkoBSW replied:
Correct, I knew that but forgot to add under things not yet working.

TomWerner commented on C7Engine/TileExtensions.cs:
```diff
@@ -40,6 +41,91 @@ public static void DisbandNonDefendingUnits(this Tile tile) {
+			}
+			return totalProgress;
+		}
+		/// <summary>
```
could you add a blank line between methods?

KulkoBSW committed:
Fixed error with multiple workers working on the same file

KulkoBSW committed:
Fixed error with multiple workers working on the same file

TomWerner commented on C7Engine/TileExtensions.cs:
```diff
@@ -40,6 +41,93 @@ public static void DisbandNonDefendingUnits(this Tile tile) {
+			if (currentWorkerJob == null) {
+				Log.Error($"can't call method FinishWorkerJob without WorkerJob");
+				return;
+			}
```
I'd leave out this null check - there's already a null check in the caller.

TomWerner commented on C7Engine/TileExtensions.cs:
```diff
@@ -40,6 +41,93 @@ public static void DisbandNonDefendingUnits(this Tile tile) {
+					continue;
+				}
+
+				if (currentWorkerJob.Equals(unit.WorkerJob)) {
```
Is there a reason to use `.Equals` instead of `==`? (Genuine question, I'm not super well versed in the nuances of C#)

KulkoBSW replied:
Okay I checked, you are right these are just my habits from other programming languages.

TomWerner commented on C7Engine/TileExtensions.cs:
```diff
@@ -40,6 +41,93 @@ public static void DisbandNonDefendingUnits(this Tile tile) {
+			if (currentWorkerJob == null) {
+				Log.Error($"can't call method FinishWorkerJob without WorkerJob");
+				return;
+			}
```
Similarly I'm not sure this null check is necessary

KulkoBSW replied:
The method is designed to be reusable, which is why I added the line. Right now a nullpointer would be thrown when I compare with a nullvalue, but that won't be the case anymore once I use ==. But still if somebody would call the function with null it would behave rather unintuitive and might create hard to find bugs which is why I thought it better to add the line. If we are concerned Efficiency/Performance I can also take it back out.

TomWerner replied:
I don't think it's a question of efficiency or performance, but more a question of why this method is special compared to all other methods - most other methods don't have explicit checks that the arguments are non-null.

You could probably get away with a comment like `// REQUIRED: currentWorkerJob != null`

TomWerner commented on C7Engine/MapUnitExtensions.cs:
```diff
@@ -527,8 +533,21 @@ public static void irrigate(this MapUnit unit) {
 			// TODO add animation and long process of building
-			unit.location.overlays.irrigation = true;
+			unit.WorkerJob = C7Action.UnitIrrigate;
+			unit.updateWorkerJob();
```
To double check my understanding here:
 - This won't cause double counting because the end of turn logic only looks at units with movement points left, and this consumes the movement points?
 - The check below ensures that if the cost was 1, we finish right away?

If both are true, it might be worth leaving a comment explaining some of the subtlety here.

TomWerner commented:
> How would I even begin to make a different behavior Multiplayer (update each Worker individually) vs. Singleplayer (Update all workers at one point and then check if the Job is done)

I wouldn't worry too much about multiplayer yet, just single player to start seems very reasonable. 

---

Overall I think this looks good! 

One other code pointer, I think we can set the button `TooltipText` field [here](https://github.com/C7-Game/Prototype/blob/Development/C7/UIElements/UnitButtons/UnitButtons.cs#L100) which would be a nice eventual improvement.

TomWerner committed:
TileAssignmentLayer: Draw the big fat cross and indicate tiles worked by other cities

#541

KulkoBSW committed:
Cleaned up NewLines,
Removed  not needed null checks,
Used == instead of Equeals for string Comp
added a few more comments.

KulkoBSW committed:
Cleaned up NewLines,
Removed  not needed null checks,
Used == instead of Equeals for string Comp
added a few more comments.

KulkoBSW committed:
Merge branch 'C7-Game:Development' into kulkobsw/multiturn-irrigation

KulkoBSW committed:
Merge branch 'C7-Game:Development' into kulkobsw/multiturn-irrigation

TomWerner committed:
Implement basic AI player research

Right now they just randomly pick techs to research without any particular strategy, but this will get us closer to some niceties, like tech trading and era advancement.

#391

TomWerner committed:
Fix tile yields for city centers

This is primarily based on the english translation of this german wiki: https://wiki.civforum.de/wiki/Stadtfeldertrag_(Civ3)

#541

TomWerner committed:
Display food/shield/commerce yields in the city screen

We can now see the result of moving citizens around.

#541

KulkoBSW committed:
Merge branch 'C7-Game:Development' into kulkobsw/multiturn-irrigation

TomWerner committed:
Run dotnet format C7.sln whitespace

TomWerner committed:
Merge pull request #564 from C7-Game/twrner/run-dotnet-format

Run `dotnet format C7.sln whitespace`

TomWerner committed:
Implement the A* path finding algorithm

This gives much nicer results when doing path finding over railroads in particular, but also should involve much less searching in the general case.

Our path finding still isn't perfect though, because it only looks at the tile movement cost, without thinking about unit movement costs. So in some cases it routes around an obstacle like jungle tiles, when it would be faster to go through them. I implemented that logic [in TilePath.cs](https://github.com/C7-Game/Prototype/blob/Development/C7GameData/AIData/TilePath.cs#L33-L68) so the goto cursor is accurate, but it hasn't been backported to the pathing logic yet.

KulkoBSW committed:
Add animation and refactor code to reuse methods.

TomWerner committed:
Delete unused pathing algorithms and their tests

TomWerner committed:
Define an EdgeWalker that is aware of unit movement costs

This avoids weird pathing where a warrior will walk the long way around a jungle, taking multiple moves to walk across grassland, because the movement cost of the jungle is 3.

TomWerner committed:
Merge pull request #566 from C7-Game/twrner/use-a-star

Implement the A* path finding algorithm

TomWerner committed:
Tune the A* land distance cost heuristic to handle railroads better

This ensures that if we have railroads in a large "C" shape, units will follow the "C", instead of taking the more direct (and expensive) path.

TomWerner committed:
Merge pull request #550 from C7-Game/twrner/change-citizen-types

Support turning citizens into specialists and changing their type

TomWerner committed:
Merge pull request #551 from C7-Game/twrner/assign-citizens-in-scenarios

 Provide tile assignments for cities in scenarios

esbudylin committed:
Revert Godot version update

esbudylin committed:
Merge pull request #530 from lt-brlara/chore/483-upgrade-to-dotnet-8

chore: :memo: upgrade target .NET framework to 8.0

esbudylin committed:
Remove LuaCiv3

The code in the LuaCiv3 projects implemented a rough prototype of
interoperating C7 code with Lua. However, it hasn't been updated in
several years and no longer works against the current C7 codebase. I
think there is no need to support it, as implementing the actual
scripting system from scratch would be easier than modifying the stale
code.

TomWerner committed:
Merge pull request #554 from C7-Game/twrner/load-culture-values

Load culture values from saves and scenarios and display them

TomWerner committed:
Merge pull request #575 from C7-Game/remove-lua-civ3

Remove LuaCiv3

TomWerner committed:
Merge pull request #556 from C7-Game/twrner/handle-culture-bounds

Properly calculate city borders based on culture

TomWerner committed:
Merge pull request #558 from C7-Game/twrner/display-bfc

TileAssignmentLayer: Draw the big fat cross and indicate tiles worked by other cities

TomWerner committed:
Merge pull request #568 from C7-Game/twrner/tune-a-star-heurstic

Tune the A* land distance cost heuristic to handle railroads better

TomWerner committed:
Merge pull request #567 from C7-Game/twrner/delete-unused-algorithms

Delete unused pathing algorithms and their tests

TomWerner committed:
Merge pull request #569 from C7-Game/twrner/unit-aware-pathing

Define an EdgeWalker that is aware of unit movement costs

TomWerner committed:
Merge pull request #561 from C7-Game/twrner/ai-researches-tech

Implement basic AI player research

TomWerner committed:
Merge pull request #562 from C7-Game/twrner/fix-tile-yields

Fix tile yields, including for city centers

TomWerner committed:
Merge pull request #565 from C7-Game/twrner/display-yields-in-city-screen

Display food/shield/commerce yields in the city screen

rchoffardet committed:
Fixing the fog of war

rchoffardet committed:
fix typo

rchoffardet committed:
fix whitespace

rchoffardet committed:
Hack to fix seams

esbudylin committed:
Load existing buildings (#580)

This commit implements loading of produced buildings from save and
scenario files. It introduces a new class, CityBuilding, which
represents a produced building, and adds functionality to store the
list of buildings within the City class.

Currently, the method for loading city buildings from Civ3 data handles
destroyed and sold buildings incorrectly, as it loads them as existing
buildings. The issue arises because Civ3 save files preserve data on both
existing and destroyed/sold buildings, and I haven't yet found a way
to distinguish between them.

rchoffardet committed:
Fix fog misalignement

TomWerner committed:
Merge pull request #582 from rchoffardet/fix-fog-of-war

Fixing the transition of the fog of war

KulkoBSW committed:
Merge remote-tracking branch 'origin/kulkobsw/multiturn-irrigation' into kulkobsw/multiturn-irrigation

# Conflicts:
#	C7Engine/EntryPoints/TurnHandling.cs
#	C7Engine/MapUnitExtensions.cs
#	C7Engine/TileExtensions.cs
